<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Удаление пропусков</a></span></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li><li><span><a href="#Удаление-пропусков-(продолжение)" data-toc-modified-id="Удаление-пропусков-(продолжение)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Удаление пропусков (продолжение)</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Проверка-гипотез" data-toc-modified-id="Проверка-гипотез-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Проверка гипотез</a></span><ul class="toc-item"><li><span><a href="#Анализ-взаимосвязи-между-количеством-детей-и-возвратом-кредита-в-срок" data-toc-modified-id="Анализ-взаимосвязи-между-количеством-детей-и-возвратом-кредита-в-срок-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Анализ взаимосвязи между количеством детей и возвратом кредита в срок</a></span></li><li><span><a href="#Анализ-взаимосвязи-между-семейным-положением-и-возвратом-кредита-в-срок" data-toc-modified-id="Анализ-взаимосвязи-между-семейным-положением-и-возвратом-кредита-в-срок-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Анализ взаимосвязи между семейным положением и возвратом кредита в срок</a></span></li><li><span><a href="#Анализ-взаимосвязи-между-уровнем-дохода-и-возвратом-кредита-в-срок" data-toc-modified-id="Анализ-взаимосвязи-между-уровнем-дохода-и-возвратом-кредита-в-срок-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Анализ взаимосвязи между уровнем дохода и возвратом кредита в срок</a></span></li><li><span><a href="#Анализ-взаимосвязи-между-целью-кредита-и-возвратом-кредита-в-срок" data-toc-modified-id="Анализ-взаимосвязи-между-целью-кредита-и-возвратом-кредита-в-срок-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Анализ взаимосвязи между целью кредита и возвратом кредита в срок</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Исследование надежности заемщиков 

**Цель исследования** — разобраться, какие факторы влияют на факт погашения кредита в срок.

**Для этого проверим 4 гипотезы**:
1. Количество детей влияет на возврат кредита в срок.
2. Семейное положение клиента связано с возвратом кредита в срок.
3. Уровень дохода влияет на вероятность погашения кредита вовремя.
4. Разные цели кредита оказывают влияние на его возврат в срок.

**Описание данных**

В этом исследовании будут использованы входные данные от банка — статистика о платёжеспособности клиентов. Мы получим их из файла `credit_worthiness_project`. Данные включают следующую информацию о клиентах банка:




* `children` — количество детей в семье


* `days_employed` — общий трудовой стаж в днях


* `dob_years` — возраст клиента в годах


* `education` — уровень образования клиента


* `education_id` — идентификатор уровня образования


* `family_status` — семейное положение


* `family_status_id` — идентификатор семейного положения


* `gender` — пол клиента


* `income_type` — тип занятости


* `debt` — имел ли задолженность по возврату кредитов


* `total_income` — ежемесячный доход в рублях


* `purpose` — цель получения кредита

**Ход исследования**

Перед проверкой гипотез потребуется обзор данных, так как отсутствует информация о качестве данных.


Проведём проверку данных на наличие ошибок и оценим их потенциальное влияние на исследование. На этапе предобработки сосредоточимся на исправлении наиболее значительных ошибок.


Таким образом, исследование пройдёт в три этапа:
 1. Обзор данных.
 
 Изучим структуру и характеристики набора данных. Посмотрим, какие проблемы могут содержать данные.
 
 
 2. Предобработка данных.
 
 Проведём более тщательную проверку на наличие ошибок и оценим их влияние. На этом этапе сосредоточимся на устранении наиболее критичных проблем, таких как пропуски, некорректные значения и дубликаты.
 
 
 3. Проверка гипотез.
 
 Используя обработанные данные, проверим сформулированные гипотезы, чтобы выявить факторы, которые влияют на возврат кредита в срок.
 

Результаты исследования будут учтены при построении модели кредитного скоринга для оценки способности потенциального заёмщика вернуть кредит банку.

## Обзор данных

Откроем таблицу и изучим общую информацию о данных:

In [1]:
# импортируем библиотеку pandas
import pandas as pd 

# считаем данные из csv-файла в датафрейм и сохраним в переменную data 
data = pd.read_csv('C:/Users/hitor/datasets/credit_worthiness_project.csv') 

In [2]:
# выведем первые 20 строчек датафрейма data на экран:
data.head(20) 

,children,days_employed,dob_years,education,education_id,family_status,family_status_id,gender,income_type,debt,total_income,purpose
0,1,-8437.673028,42,высшее,0,женат / замужем,0,F,сотрудник,0,253875.639453,покупка жилья
1,1,-4024.803754,36,среднее,1,женат / замужем,0,F,сотрудник,0,112080.014102,приобретение автомобиля
2,0,-5623.422610,33,Среднее,1,женат / замужем,0,M,сотрудник,0,145885.952297,покупка жилья
3,3,-4124.747207,32,среднее,1,женат / замужем,0,M,сотрудник,0,267628.550329,дополнительное образование
4,0,340266.072047,53,среднее,1,гражданский брак,1,F,пенсионер,0,158616.077870,сыграть свадьбу
5,0,-926.185831,27,высшее,0,гражданский брак,1,M,компаньон,0,255763.565419,покупка жилья
6,0,-2879.202052,43,высшее,0,женат / замужем,0,F,компаньон,0,240525.971920,операции с жильем
7,0,-152.779569,50,СРЕДНЕЕ,1,женат / замужем,0,M,сотрудник,0,135823.934197,образование
8,2,-6929.865299,35,ВЫСШЕЕ,0,гражданский брак,1,F,сотрудник,0,95856.832424,на проведение свадьбы
9,0,-2188.756445,41,среднее,1,женат / замужем,0,M,сотрудник,0,144425.938277,покупка жилья для семьи


Предварительно в таблице с данными можно заметить следующие проблемы:
* отрицательное количество дней трудового стажа в столбце `days_employed`.
* неявные дубликаты в столбце `education`. В нём есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв.
* отсутствуют категории для значений в столбцах `total_income`, `purpose` и `dob_years`. Это может затруднить анализ данных.

In [3]:
# выведем основную информацию о датафрейме с помощью метода info():
data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Во всех столбцах таблицы кроме `days_employed` и `total_income` одинаковое количество значений. Значит, эти столбцы содержат пропуски.

**Выводы**

* В каждом столбце таблицы представлена информация о клиентах: количество детей в семье, общий трудовой стаж, возраст, уровень образования и идентификатор уровня, семейный статус и идентификатор статуса, пол, вид деятельности, наличие задолженности, ежемесячный доход и цель кредита.


* На данном этапе можно предположить, что данных достаточно для проверки выдвинутых гипотез. Однако в столбцах `days_employed` и `total_income` замечены пропуски. Также была замечена аномалия в столбце `days_employed` - отрицательное количество дней трудового стажа, неявные дубликаты в столбце `education`. Отсутствуют категории для данных о ежемесячном доходе, цели кредита, возрасте клиентов. Без категорий сложно изучать различия и искать закономерности в данных.

Устраним проблемы в данных, чтобы двигаться дальше.

## Предобработка данных

### Удаление пропусков

In [4]:
# выведем количество пропущенных значений для каждого столбца, используем комбинацию двух методов
data.isna().sum()  

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце обработаем на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости сотрудник пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.

In [5]:
# найдём медиану дохода по каждому типу деятельности из столбца income_type
# заполним пропуски медианой в столбце total_income и сохраним результат
data['total_income'] = data['total_income'].fillna(data.groupby('income_type')['total_income'].transform('median')) 

# убедимся, что пропуски заполнены 
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

### Обработка аномальных значений

Обработаем значения в столбце `days_employed`: заменим все отрицательные значения положительными: 

In [6]:
# заменим значения с помощью метода abs() и сохраним результат 
data['days_employed'] = data['days_employed'].abs() 

# выведем первые 20 строк датафрейма и проверим результат
data.head(10) 

,children,days_employed,dob_years,education,education_id,family_status,family_status_id,gender,income_type,debt,total_income,purpose
0,1,8437.673028,42,высшее,0,женат / замужем,0,F,сотрудник,0,253875.639453,покупка жилья
1,1,4024.803754,36,среднее,1,женат / замужем,0,F,сотрудник,0,112080.014102,приобретение автомобиля
2,0,5623.422610,33,Среднее,1,женат / замужем,0,M,сотрудник,0,145885.952297,покупка жилья
3,3,4124.747207,32,среднее,1,женат / замужем,0,M,сотрудник,0,267628.550329,дополнительное образование
4,0,340266.072047,53,среднее,1,гражданский брак,1,F,пенсионер,0,158616.077870,сыграть свадьбу
5,0,926.185831,27,высшее,0,гражданский брак,1,M,компаньон,0,255763.565419,покупка жилья
6,0,2879.202052,43,высшее,0,женат / замужем,0,F,компаньон,0,240525.971920,операции с жильем
7,0,152.779569,50,СРЕДНЕЕ,1,женат / замужем,0,M,сотрудник,0,135823.934197,образование
8,2,6929.865299,35,ВЫСШЕЕ,0,гражданский брак,1,F,сотрудник,0,95856.832424,на проведение свадьбы
9,0,2188.756445,41,среднее,1,женат / замужем,0,M,сотрудник,0,144425.938277,покупка жилья для семьи


Для каждого типа занятости выведем медианное значение трудового стажа `days_employed` в днях:

In [7]:
# найдём медиану трудового стажа по каждому типу деятельности из столбца income_type
data.groupby('income_type')['days_employed'].median() 

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть. Тем более этот столбец не понадобится нам для исследования.

Проверим столбцы `children`, `debt` и `family_status` на наличие аномальных значений.

Выведем перечень уникальных значений столбца `children`:

In [8]:
# используем метод unique() для получения уникальных значений
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

В столбце `children` есть два аномальных значения: -1 и 20. Удалим строки, в которых встречаются такие аномальные значения из датафрейма data.

In [9]:
# удалим аномальные значения и сохраним результат в data
data = data[(data['children'] != -1) & (data['children'] != 20)]

# ещё раз выведем список уникальных значений для проверки результата
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

Выведем на экран перечень уникальных значений столбцов `debt` и `family_status`:

In [10]:
# используем метод unique() для получения уникальных значений
data['debt'].unique()

array([0, 1], dtype=int64)

In [11]:
# используем метод unique() для получения уникальных значений
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

Убедились, что в столбцах `debt` и `family_status` нет аномальных значений.

### Удаление пропусков (продолжение)

Заполним пропуски в столбце `days_employed` медианными значениями по каждому типу занятости `income_type`:

In [12]:
# найдём медиану трудового стажа по каждому типу деятельности из столбца income_type
# заполним пропуски медианой в столбце days_employed и сохраним результат
data['days_employed'] = data['days_employed'].fillna(data.groupby('income_type')['days_employed'].transform('median')) 

# убедимся, что пропуски заполнены 
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

В этом исследовании сосредоточимся на доходе в рублях без копеек. Целые числа легче интерпретировать и сравнивать, поэтому заменим вещественный тип данных в столбце `total_income` на целочисленный: 

In [13]:
# произведём замену с помощью метода astype() и сохраним результат
data['total_income'] = data['total_income'].astype(int)

# проверим результат
data.info() 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21402 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21402 non-null  int64  
 1   days_employed     21402 non-null  float64
 2   dob_years         21402 non-null  int64  
 3   education         21402 non-null  object 
 4   education_id      21402 non-null  int64  
 5   family_status     21402 non-null  object 
 6   family_status_id  21402 non-null  int64  
 7   gender            21402 non-null  object 
 8   income_type       21402 non-null  object 
 9   debt              21402 non-null  int64  
 10  total_income      21402 non-null  int32  
 11  purpose           21402 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(5)
memory usage: 2.0+ MB


### Обработка дубликатов

Обработаем неявные дубликаты в столбце `education`:

In [14]:
# приведём значения к нижнему регистру с помощью метода str.lower()
data['education'] = data['education'].str.lower()

# выведем 20 строк датафрейма и проверим результат
data.head(20)

,children,days_employed,dob_years,education,education_id,family_status,family_status_id,gender,income_type,debt,total_income,purpose
0,1,8437.673028,42,высшее,0,женат / замужем,0,F,сотрудник,0,253875,покупка жилья
1,1,4024.803754,36,среднее,1,женат / замужем,0,F,сотрудник,0,112080,приобретение автомобиля
2,0,5623.422610,33,среднее,1,женат / замужем,0,M,сотрудник,0,145885,покупка жилья
3,3,4124.747207,32,среднее,1,женат / замужем,0,M,сотрудник,0,267628,дополнительное образование
4,0,340266.072047,53,среднее,1,гражданский брак,1,F,пенсионер,0,158616,сыграть свадьбу
5,0,926.185831,27,высшее,0,гражданский брак,1,M,компаньон,0,255763,покупка жилья
6,0,2879.202052,43,высшее,0,женат / замужем,0,F,компаньон,0,240525,операции с жильем
7,0,152.779569,50,среднее,1,женат / замужем,0,M,сотрудник,0,135823,образование
8,2,6929.865299,35,высшее,0,гражданский брак,1,F,сотрудник,0,95856,на проведение свадьбы
9,0,2188.756445,41,среднее,1,женат / замужем,0,M,сотрудник,0,144425,покупка жилья для семьи


Проверим остальные столбцы на наличие дубликатов:

In [15]:
# выведем на экран количество строк-дубликатов в данных
data.duplicated().sum()

71

Удалим строки-дубликаты: 

In [16]:
# для удаления используем метод drop_duplicates()
data = data.drop_duplicates()

# проверим результат
data.duplicated().sum()

0

### Категоризация данных

На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


Например, кредитополучателю с доходом 25000 назначим категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используем собственную функцию с именем `categorize_income()` и метод `apply()`.

In [17]:
# напишем функцию для разделения данных на категории с параметром - столбец с ежемесячным доходом клиента
# в теле функции используем условные конструкции для разделения данных
def categorize_income(total_income): 
    
    if 0 <= total_income <= 30000:
        return 'E'
    if 30001 <= total_income <= 50000:
        return 'D'
    if 50001 <= total_income <= 200000:
        return 'C'
    if 200001 <= total_income <= 1000000:
        return 'B'
    return 'A'

# выведем результат, чтобы протестировать функцию
# print(categorize_income(50005))

In [18]:
# применим функцию методом apply() и сохраним результат в столбец total_income_category
data['total_income_category'] = data['total_income'].apply(categorize_income) 

# проверим результат
data.head()

,children,days_employed,dob_years,education,education_id,family_status,family_status_id,gender,income_type,debt,total_income,purpose,total_income_category
0,1,8437.673028,42,высшее,0,женат / замужем,0,F,сотрудник,0,253875,покупка жилья,B
1,1,4024.803754,36,среднее,1,женат / замужем,0,F,сотрудник,0,112080,приобретение автомобиля,C
2,0,5623.422610,33,среднее,1,женат / замужем,0,M,сотрудник,0,145885,покупка жилья,C
3,3,4124.747207,32,среднее,1,женат / замужем,0,M,сотрудник,0,267628,дополнительное образование,B
4,0,340266.072047,53,среднее,1,гражданский брак,1,F,пенсионер,0,158616,сыграть свадьбу,C


Создадим в датафрейме `data` столбец `age_category` с категориями на основании следующих диапазонов:

- 18–25 — `от 18 до 25 лет`;
- 25-30 — `от 25 до 30 лет`;
- 30-40 — `от 30 до 40 лет`;
- 40-50 — `от 40 до 50 лет`;
- 50 и выше — `от 50 лет`.


Например, клиенту в возрасте 20 лет назначим категорию ``от 18 до 25 лет``. Напишем собственную функцию с именем `categorize_age` и используем метод `apply()`.

In [19]:
# напишем функцию для разделения данных на категории с параметром - столбец с возрастом клиента
# в теле функции используем условные конструкции для разделения данных
def categorize_age(age):
    
    if 18 <= age <= 25:
        return 'от 18 до 25 лет'
    elif 25 <= age <= 30:
        return 'от 25 до 30 лет'
    elif 30 <= age <= 40:
        return 'от 30 до 40 лет'
    elif 40 <= age <= 50:
        return 'от 40 до 50 лет'
    return 'от 50 лет'

# выведем результат, чтобы протестировать функцию
# print(categorize_age(19))

In [20]:
# применим функцию методом apply() и сохраним результат в столбец age_category
data['age_category'] = data['dob_years'].apply(categorize_age)

# проверим результат
data.head()

,children,days_employed,dob_years,education,education_id,family_status,family_status_id,gender,income_type,debt,total_income,purpose,total_income_category,age_category
0,1,8437.673028,42,высшее,0,женат / замужем,0,F,сотрудник,0,253875,покупка жилья,B,от 40 до 50 лет
1,1,4024.803754,36,среднее,1,женат / замужем,0,F,сотрудник,0,112080,приобретение автомобиля,C,от 30 до 40 лет
2,0,5623.422610,33,среднее,1,женат / замужем,0,M,сотрудник,0,145885,покупка жилья,C,от 30 до 40 лет
3,3,4124.747207,32,среднее,1,женат / замужем,0,M,сотрудник,0,267628,дополнительное образование,B,от 30 до 40 лет
4,0,340266.072047,53,среднее,1,гражданский брак,1,F,пенсионер,0,158616,сыграть свадьбу,C,от 50 лет


Перед тем, как разделить данные на категории по целям кредита, выведем на экран перечень уникальных целей взятия кредита из столбца `purpose`:

In [21]:
# выведем список уникальных значений с помощью метода unique()
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:

- `'операции с автомобилем'`;
- `'операции с недвижимостью'`;
- `'проведение свадьбы'`;
- `'получение образования'`.

Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.

Используем собственную функцию с именем `categorize_purpose()` и метод `apply()`:

In [22]:
# напишем функцию для разделения данных на категории с параметром - строка в датафрейме
# в теле функции используем условные конструкции для разделения данных
def categorize_purpose(row):
    if 'автом' in row:
        return 'операции с автомобилем'
    elif 'жил' in row or 'недвиж' in row:
        return 'операции с недвижимостью'
    elif 'свад' in row:
        return 'проведение свадьбы'
    elif 'образов' in row:
        return 'получение образования'
    return 'нет категории'

# выведем результат, чтобы протестировать функцию
# print(categorize_purpose('свадьба')) 

In [23]:
# применим функцию методом apply() и сохраним результат в столбец purpose_category
data['purpose_category'] = data['purpose'].apply(categorize_purpose) 

# проверим результат
data.head(10)

,children,days_employed,dob_years,education,education_id,family_status,family_status_id,gender,income_type,debt,total_income,purpose,total_income_category,age_category,purpose_category
0,1,8437.673028,42,высшее,0,женат / замужем,0,F,сотрудник,0,253875,покупка жилья,B,от 40 до 50 лет,операции с недвижимостью
1,1,4024.803754,36,среднее,1,женат / замужем,0,F,сотрудник,0,112080,приобретение автомобиля,C,от 30 до 40 лет,операции с автомобилем
2,0,5623.422610,33,среднее,1,женат / замужем,0,M,сотрудник,0,145885,покупка жилья,C,от 30 до 40 лет,операции с недвижимостью
3,3,4124.747207,32,среднее,1,женат / замужем,0,M,сотрудник,0,267628,дополнительное образование,B,от 30 до 40 лет,получение образования
4,0,340266.072047,53,среднее,1,гражданский брак,1,F,пенсионер,0,158616,сыграть свадьбу,C,от 50 лет,проведение свадьбы
5,0,926.185831,27,высшее,0,гражданский брак,1,M,компаньон,0,255763,покупка жилья,B,от 25 до 30 лет,операции с недвижимостью
6,0,2879.202052,43,высшее,0,женат / замужем,0,F,компаньон,0,240525,операции с жильем,B,от 40 до 50 лет,операции с недвижимостью
7,0,152.779569,50,среднее,1,женат / замужем,0,M,сотрудник,0,135823,образование,C,от 40 до 50 лет,получение образования
8,2,6929.865299,35,высшее,0,гражданский брак,1,F,сотрудник,0,95856,на проведение свадьбы,C,от 30 до 40 лет,проведение свадьбы
9,0,2188.756445,41,среднее,1,женат / замужем,0,M,сотрудник,0,144425,покупка жилья для семьи,C,от 40 до 50 лет,операции с недвижимостью


**Выводы**

Предобработка обнаружила следующие проблемы в данных:

- Пропуски в столбцах `days_employed` и `total_income`;
- Аномалии в столбце `days_employed` в виде отрицательных значений, а также 2 артефакта в столбце `children`;
- Вещественный тип данных в столбце `total_income` вместо целочисленного;
- Явные и неявные дубликаты в данных;
- Отсутствие категорий для информации в столбцах `total_income`, `dob_years` и `purpose`.

Пропуски в столбцах `days_employed` и `total_income` могли возникнуть по техническим причинам. Для принятия решения о выдаче кредита банк оценивает трудовой стаж и уровень дохода клиента. Значит, клиенту обязательно нужно указать эти данные и, скорее всего, он не оставлял эти поля пустыми при заполнении анкеты. Пропуски в столбцах с информацией о заработной плате и трудовом стаже в днях были заполнены медианой. В случае с количественными переменными лучше заполнить пропуски медианным значением, так как в данных возможны значения, которые значительно отличаются от других и поэтому могут сместить среднее. 

Отрицательные значения в столбце `days_employed` были заменены на положительные. Устранены артефакты в стобце `children`.

Вещественный тип данных в столбце `total_income` заменили на целочисленный. Удалили дубликаты в данных. Разбили данные в столбцах `total_income`, `dob_years` и `purpose` на категории и сохранили в новых столбцах `total_income_category`, `age_category` и `purpose_category`.

Теперь можно перейти к проверке гипотез.

## Проверка гипотез

###  Анализ взаимосвязи между количеством детей и возвратом кредита в срок

Чтобы проверить, влияет ли количество детей клиента на факт погашения кредита в срок: 

* Посмотрим, сколько клиентов входит в каждую категорию по количеству детей.
* Проверим, сколько клиентов в каждой группе имеют задолженность по возврату кредитов.
* Выясним, какую долю от всех клиентов составляют клиенты с задолженностью в каждой категории.

In [24]:
# используем методы groupby() и agg() вместе, чтобы одновременно применить две функции к столбцу 'children'
# выведем на экран общее количество клиентов и количество клиентов с задолженностью в каждой категории столбца 'children'
data.groupby('children').agg({'debt':['count', 'sum']})

debt      
          count   sum
children             
0         14091  1063
1          4808   444
2          2052   194
3           330    27
4            41     4
5             9     0

Наибольшее число клиентов, представленных в таблице, `не имеют детей`. Далее следуют категории `с одним ребенком` и `двумя детьми`. Эти же категории охватывают самое большое число клиентов с задолженностью.

Самая маленькая группа состоит из клиентов с количеством детей - 5. Согласно данным, все клиенты в категории `"5"` возвращают кредит в срок.

Чтобы найти долю клиентов с задолженностью рассчитаем конверсию. Разделим количество клиентов с задолженностью на общее количество клиентов во всех категориях:

In [25]:
# объявим функцию для рассчёта конверсии с параметрами: таблица с данными, столбцы категории клиентов и залдолженности 
def conversion(data, category, debt): 

# сгруппируем данные по столбцу с категорией и применим к нему агрегирующие функции
# рассчитаем долю клиентов с задолженностью и округлим результат
# вернём результат, отсортированный в порядке убывания
    category_count_sum = data.groupby(category)[debt].agg(['count', 'sum'])
    category_count_sum['conversion'] = (category_count_sum['sum'] / category_count_sum['count'] * 100).round(1)
    return category_count_sum['conversion'].sort_values(ascending=False)

In [26]:
# вызовем функцию, передадим ей таблицу с данными, столбцы 'children' и 'debt' в качестве аргументов
# выведем результат на экран
children_conversion = conversion(data, 'children', 'debt')
children_conversion

children
4    9.8
2    9.5
1    9.2
3    8.2
0    7.5
5    0.0
Name: conversion, dtype: float64

Соберём полученные данные в таблицу для более комфортного восприятия:

In [27]:
# создадим таблицу с помощью конструктора библиотеки pandas
# сохраним таблицу в переменную children_conversion_table
children_conversion_table = pd.DataFrame(  
    data=[
        [4, 41, 4, 9.8], 
        [2, 2052, 194, 9.5], 
        [1, 4808, 444, 9.2], 
        [3, 330, 27, 8.2], 
        [0, 14091, 1063, 7.5],
        [5, 9, 0, 0.0]
        
    ],
    columns = ['children_category', 'customers_number', 'debt', 'conversion_%']
)

# выведем результат на экран
children_conversion_table

,children_category,customers_number,debt,conversion_%
0,4,41,4,9.8
1,2,2052,194,9.5
2,1,4808,444,9.2
3,3,330,27,8.2
4,0,14091,1063,7.5
5,5,9,0,0.0


**Вывод:** 

В результате расчетов выяснили:
1. Большинство клиентов с задолженностью входит в категорию `c 4 детьми` - **9.8%**. В ней задолженность имеет каждый десятый клиент.
2. Близкое значение имеют группы: `с двумя детьми` - **9.5%**, `с одним ребёнком` - **9.2%**.
2. Меньше доля клиентов с задолженностью в категориях `с тремя детьми` и `без детей` - **8.2%** и **7.5%**.
3. Клиенты `с 5 детьми` **не имеют задолженностей** по кредиту. В эту категорию входит самое маленькое количество клиентов - всего лишь 9 человек.

Данные показывают, что нельзя утверждать о прямой зависимости между количеством детей и возвратом кредита в срок, так как у клиентов `с 1 ребёнком` задолженность по кредиту на **1%** выше, чем у клиентов `с тремя детьми`. С другой стороны, клиенты `без детей` имеют самый низкий процент задолженности. `Пятая категория` включает слишком маленькое количество клиентов для корректного вывода. Таким образом, есть слабая зависимость между количеством детей и погашением кредита в срок.

 ###  Анализ взаимосвязи между семейным положением и возвратом кредита в срок

Проверим, влияет ли семейное положение на факт погашения кредита в срок:

* Выясним, сколько клиентов охватывает каждая категория по семейному положению.
* Посмотрим, сколько клиентов в каждой категории не оплачивает кредит в срок.
* Рассчитаем долю клиентов с задолженностью в зависимости от семейного статуса.

In [28]:
# используем методы groupby() и agg(), чтобы применить агрегирующие функции к столбцу 'family_status'
# отсортируем по убыванию значения в столбцах count и sum
# выведем на экран общее количество клиентов и количество клиентов с задолженностью в каждой категории столбца 'family_status'
data.groupby('family_status').agg({'debt':['count', 'sum']}).sort_values(by=[('debt','count'),('debt','sum')], ascending=False)

debt     
                       count  sum
family_status                    
женат / замужем        12261  927
гражданский брак        4134  385
Не женат / не замужем   2796  273
в разводе               1189   84
вдовец / вдова           951   63

Самую многочисленную группу по количеству клиентов в целом и количеству клиентов с задолженностью составляют клиенты `в браке`, а самую маленькую - `"вдовец / вдова"`.

In [29]:
# вызовем функцию для расчёта конверсии, передадим ей таблицу с данными, столбцы 'family_status' и 'debt'
# выведем результат на экран
family_status_conversion = conversion(data, 'family_status', 'debt')
family_status_conversion

family_status
Не женат / не замужем    9.8
гражданский брак         9.3
женат / замужем          7.6
в разводе                7.1
вдовец / вдова           6.6
Name: conversion, dtype: float64

Соберём данные в таблицу:

In [30]:
# создадим таблицу и сохраним её в переменную family_conversion_table
family_conversion_table = pd.DataFrame(  
    data=[
        ['не женат / не замужем', 2796, 273, 9.8], 
        ['гражданский брак', 4134, 385, 9.3], 
        ['женат / замужем', 12261, 927, 7.6], 
        ['в разводе', 1189, 84, 7.1], 
        ['вдовец / вдова', 951, 63, 6.6]
        
    ],
    columns = ['family_status_category', 'customers_number', 'debt', 'conversion_%']
)

# выведем результат на экран
family_conversion_table

,family_status_category,customers_number,debt,conversion_%
0,не женат / не замужем,2796,273,9.8
1,гражданский брак,4134,385,9.3
2,женат / замужем,12261,927,7.6
3,в разводе,1189,84,7.1
4,вдовец / вдова,951,63,6.6


**Вывод:** 

В ходе исследования установили:

1. Самыми крупными категориями по количеству клиентов с задолженностью являются `"не женат / не замужем"` - **9.8%** и `"гражданский брак"` - **9.3%**. Каждый десятый клиент в этой категории имеет задолженность по кредиту.
2. Клиенты `в браке` реже имеют задолженность - **7.6%**, несмотря на самое большое количество клиентов в целом и клиентов с задолженностью в сравнении с другими категориями. Далее следуют `клиенты в разводе` - **7.1%**.
3. `Вдовцы и вдовы` имеют самый низкий процент задолженности - **6.6%**.

Прослеживается зависимость между семейным положением клиента и своевременным погашением кредита. Особенно заметна разница в **3%** между долями в категориях клиентов `"не женат / не замужем"` и `"вдовец / вдова"`.

###  Анализ взаимосвязи между уровнем дохода и возвратом кредита в срок

Проверим, влияет ли уровень дохода на возврат кредита в срок:

* Выясним, сколько клиентов охватывает каждая категория по уровню дохода.
* Посмотрим, сколько клиентов с задолженностью есть в каждой категории.
* Ознакомимся с видом деятельности клиентов в каждой категории дохода.
* Ознакомимся с информацией о клиентах в каждой категории видов деятельности.
* Рассчитаем долю клиентов с задолженностью в каждой категории дохода.

In [31]:
# применим функции count и sum к столбцу 'total_income_category' 
# отсортируем по убыванию значения в столбцах count и sum
# выведем на экран общее количество клиентов и количество клиентов с задолженностью в каждой категории столбца 'total_income_category'
data.groupby('total_income_category').agg({'debt': ['count', 'sum']}).sort_values(by=[('debt','count'), ('debt','sum')], ascending=False) 

debt      
                       count   sum
total_income_category             
C                      15921  1353
B                       5014   354
D                        349    21
A                         25     2
E                         22     2

Самая крупная категория по количеству клиентов в целом и клиентов с задолженностью - `"C"` с ежемесячным доходом от 50001 до 200000. Самые маленькие группы с небольшой разницей в количестве клиентов  - `"E"` с доходом 0-30 000, и `"A"` с доходом от 1 000 001.

Посмотрим, какие виды деятельности входят в каждую категорию дохода, какое количество клиентов в целом и количество клиентов с задолженностью охватывает каждый вид деятельности:

In [32]:
# сгруппируем данные по столбцам 'total_income_category' и 'income_type'
# применим count к стобцу 'income_type', сохраним результат в переменную income_type_count
# применим sum к стобцу 'debt', сохраним результат в переменную income_type_sum
# результат запишем в переменную income_type_count_sum и выведем на экран
income_type_count_sum = data.groupby(['total_income_category', 'income_type']).agg(
    income_type_count=('income_type', 'count'),
    income_type_sum=('debt','sum')
)

income_type_count_sum

income_type_count  income_type_sum
total_income_category income_type                                        
A                     компаньон                       15                2
                      сотрудник                       10                0
B                     безработный                      1                0
                      госслужащий                    366               18
                      компаньон                     1733              109
                      пенсионер                      552               31
                      предприниматель                  2                0
                      сотрудник                     2360              196
C                     безработный                      1                1
                      в декрете                        1                1
                      госслужащий                   1060               67
                      компаньон                     3273              262
                      пенсионер                     3081              176
                      сотрудник                     8504              846
                      студент                          1                0
D                     госслужащий                     24                1
                      компаньон                       25                1
                      пенсионер                      164                7
                      сотрудник                      136               12
E                     госслужащий                      1                0
                      компаньон                        1                0
                      пенсионер                       15                2
                      сотрудник                        5                0

Больше всего видов деятельности охватывает группа `"C"` c семью видами. Группа `"A"`, наоборот, охватывает всего 2 вида деятельности: `"компаньон"` и `"сотрудник"`. Вид деятельности `"сотрудник"` преобладает в категориях `'B'` и `'C'` по количеству сотрудников в целом и сотрудников с задолженностью. В категориях `'D'` и `'E'` преобладают пенсионеры.

Посмотрим, какую долю задолженности имеют клиенты в разных категориях видов деятельности:

In [33]:
# расчитаем конверсию: разделим количество клиентов с задолженностью на количество клиентов в целом, умножим на 100
# округлим полученные значения
# создадим столбец 'conversion_%' и сохраним в него результат
income_type_count_sum['conversion_%'] = round((income_type_count_sum['income_type_sum'] / income_type_count_sum['income_type_count'] * 100), 1)
income_type_count_sum

income_type_count  income_type_sum  \
total_income_category income_type                                           
A                     компаньон                       15                2   
                      сотрудник                       10                0   
B                     безработный                      1                0   
                      госслужащий                    366               18   
                      компаньон                     1733              109   
                      пенсионер                      552               31   
                      предприниматель                  2                0   
                      сотрудник                     2360              196   
C                     безработный                      1                1   
                      в декрете                        1                1   
                      госслужащий                   1060               67   
                      компаньон                     3273              262   
                      пенсионер                     3081              176   
                      сотрудник                     8504              846   
                      студент                          1                0   
D                     госслужащий                     24                1   
                      компаньон                       25                1   
                      пенсионер                      164                7   
                      сотрудник                      136               12   
E                     госслужащий                      1                0   
                      компаньон                        1                0   
                      пенсионер                       15                2   
                      сотрудник                        5                0   

                                       conversion_%  
total_income_category income_type                    
A                     компаньон                13.3  
                      сотрудник                 0.0  
B                     безработный               0.0  
                      госслужащий               4.9  
                      компаньон                 6.3  
                      пенсионер                 5.6  
                      предприниматель           0.0  
                      сотрудник                 8.3  
C                     безработный             100.0  
                      в декрете               100.0  
                      госслужащий               6.3  
                      компаньон                 8.0  
                      пенсионер                 5.7  
                      сотрудник                 9.9  
                      студент                   0.0  
D                     госслужащий               4.2  
                      компаньон                 4.0  
                      пенсионер                 4.3  
                      сотрудник                 8.8  
E                     госслужащий               0.0  
                      компаньон                 0.0  
                      пенсионер                13.3  
                      сотрудник                 0.0

В категориях `"B"` и `"D"` самая высокая доля задолженности у `сотрудников`. В категориях `"A"` и `"E"` представлено слишком маленькое количество клиентов для корректного вывода. В категории `"C"` всего один клиент является `безработным` и всего один находится `в декрете`, поэтому данные показывают у этих групп конверсию - 100%.

Рассчитаем долю клиентов с задолженностью в каждой категории дохода:

In [34]:
# вызовем функцию для расчёта конверсии, передадим ей таблицу с данными, столбцы 'total_income_category' и 'debt'
# выведем результат на экран
total_income_conversion = conversion(data, 'total_income_category', 'debt')
total_income_conversion

total_income_category
E    9.1
C    8.5
A    8.0
B    7.1
D    6.0
Name: conversion, dtype: float64

Создадим таблицу с результатами:

In [35]:
# создадим таблицу и сохраним её в переменную income_conversion_table
income_conversion_table = pd.DataFrame(  
    data=[
        ["E: 0-30 000", 22, 2, 9.1], 
        ["C: 50 001-200 000", 15921, 1353, 8.5], 
        ["A: от 1 000 001", 25, 2, 8.0], 
        ["B: 200 001-1 000 000", 5014, 354, 7.1], 
        ["D: 30 001-50 000", 349, 21, 6.0]
        
    ],
    columns = ['total_income_month', 'customers_number', 'debt', 'conversion_%']
)

# выведем результат на экран
income_conversion_table

,total_income_month,customers_number,debt,conversion_%
0,E: 0-30 000,22,2,9.1
1,C: 50 001-200 000,15921,1353,8.5
2,A: от 1 000 001,25,2,8.0
3,B: 200 001-1 000 000,5014,354,7.1
4,D: 30 001-50 000,349,21,6.0


**Вывод:** 

Данные показывают:

1. Данные показывают, что самый высокий процент задолженности имеет категория клиентов с общим доходом `от 0-30 000` - **9.1%**. Однако эту информацию нельзя считать достоверной, так как в категории `"E"` представлено слишком маленькое количество клиентов.
2. Меньше процент задолженности у категорий клиентов с доходом `50 001-200 000` - **8.5%**. Это самая большая группа по количеству клиентов с разными видами деятельности, самые крупные из которых : `"сотрудник"` - 8504 человек, `"компаньон"` - 3273 и `"пенсионер"` - 3081.
3. Далее следует категория с доходом `от 1 000 001` и выше с долей **8.0%**. Трудно утверждать о корректности вывода, так как в категории `"А"` представлено слишком маленькое количество клиентов. Она охватывает всего два вида деятельности. В этой категории только `компаньоны` имеют задолженность. 
4. Самую маленькую долю составляют клиенты с доходом `200 001-1 000 000` - **7.1%** и `30 001-50 000` - **6.0%**. В категориях `"B"` и `"D"` самую высокую долю задолженности имеют `сотрудники`.

### Анализ взаимосвязи между целью кредита и возвратом кредита в срок

Выясним, как разные цели кредита влияют на его оплату в срок:

* Найдём общее количество клиентов и количество клиентов с задолженностью в каждой категории по цели кредита.
* Рассчитаем долю клиентов с задолженностью.
* Для каждой цели кредита выявим долю клиентов с задолженностью в зависимости от уровня дохода, семейного положения и возраста клиентов.

In [36]:
# применим функции count и sum к столбцу 'purpose_category' и ознакомимся с результатом
data.groupby('purpose_category').agg({'debt': ['count', 'sum']}).sort_values(by=[('debt','count'), ('debt','sum')], ascending=False) 

debt     
                          count  sum
purpose_category                    
операции с недвижимостью  10751  780
операции с автомобилем     4279  400
получение образования      3988  369
проведение свадьбы         2313  183

Самая большая группа по количеству клиентов в целом и количеству клиентов с задолженностью - `"операции с недвижимостью"`, самая маленькая группа - `"проведение свадьбы"`.	

Выявим долю клиентов с задолженностью для каждой цели кредита:

In [37]:
# вызовем функцию для расчёта конверсии, передадим ей таблицу с данными, столбцы 'purpose_category' и 'debt'
# выведем результат на экран
purpose_category_conversion = conversion(data, 'purpose_category', 'debt')
purpose_category_conversion

purpose_category
операции с автомобилем      9.3
получение образования       9.3
проведение свадьбы          7.9
операции с недвижимостью    7.3
Name: conversion, dtype: float64

Представим результат в виде таблицы:

In [38]:
# создадим таблицу и сохраним её в переменную purpose_conversion_table
purpose_conversion_table = pd.DataFrame(  
    data=[
        ["операции с автомобилем ", 4279, 400, 9.3], 
        ["получение образования ", 3988, 369, 9.3], 
        ["проведение свадьбы", 2313, 183, 7.9], 
        ["операции с недвижимостью", 10751, 780, 7.3]
        
    ],
    columns = ['purpose_category', 'customers_number', 'debt', 'conversion_%']
)

# выведем результат на экран
purpose_conversion_table

,purpose_category,customers_number,debt,conversion_%
0,операции с автомобилем,4279,400,9.3
1,получение образования,3988,369,9.3
2,проведение свадьбы,2313,183,7.9
3,операции с недвижимостью,10751,780,7.3


**Посмотрим, как разные цели кредита влияют на его возврат в срок. Для этого рассчитаем долю клиентов с задолженностью для каждой цели кредита:**

* `операции с автомобилем`
* `получение образования`
* `проведение свадьбы`
* `операции с недвижимостью`

**В зависимости от категорий**:

* `уровень дохода`

* `семейное положение`

* `возраст клиентов`

Исключили категорию `количество детей` клиента, так как выявили, что она практически не влияет на возврат кредита в срок.

Рассчитаем долю клиентов с задолженностью с целью `"операции с автомобилем"` в зависимости от `ежемесячного дохода`:

In [39]:
# напишем функцию для рассчёта конверсии с параметрами: таблица с данными, столбец с целью кредита, столбец с категорией
# выберем строки с целью кредита в столбце 'purpose_category' и отфильтруем данные, сохраним результат
# сгруппируем отфильтрованные данные по столбцам с целью кредита и категорией
# посчитаем количество клиентов и количество клиентов с задолженностью  
# сбросим индекс
def category_conversion(data, purpose_category, category_column): 
    purpose_category_data = data[data['purpose_category'] == purpose_category]
    purpose_count_sum = purpose_category_data.groupby(['purpose_category', category_column]).agg(
    customer_count=(category_column,'count'),
    debt_sum=('debt','sum')
    ).reset_index()

# рассчитаем конверсию, округлим результат и добавим новый столбец в таблицу
# отсортируем значения по убыванию и заново пронумеруем строки
# вернём таблицу с результатом

    purpose_count_sum['conversion_%'] = round(
        (purpose_count_sum['debt_sum'] / purpose_count_sum['customer_count'] * 100), 1
    )
    purpose_count_sum = purpose_count_sum.sort_values(by='conversion_%', ascending=False).reset_index(drop=True)
    return purpose_count_sum

In [40]:
# выведем на экран результат для цели "операции с автомобилем" и "категории ежемесячного дохода"
category_conversion(data, 'операции с автомобилем', 'total_income_category')

,purpose_category,total_income_category,customer_count,debt_sum,conversion_%
0,операции с автомобилем,D,72,10,13.9
1,операции с автомобилем,C,3181,305,9.6
2,операции с автомобилем,B,1022,85,8.3
3,операции с автомобилем,A,2,0,0.0
4,операции с автомобилем,E,2,0,0.0


Самая высокая доля клиентов с задолженностью с целью `"операции с автомобилем"` в зависимости от `ежемесячного дохода` находится в категории `"D"`-  13.9%. Клиенты в этой категории имеют доход `от 30 001 до 50 000`. Клиенты с целью операций с автомобилем в категориях дохода `"A" - от 1 000 001` и `"E" - 0-30 000` не имеют задолженностей.

Рассчитаем долю клиентов с задолженностью с целью `"операции с автомобилем"` в зависимости от `семейного положения`:

In [41]:
# выведем на экран результат для цели "операции с автомобилем" и "категории семейного положения"
category_conversion(data, 'операции с автомобилем', 'family_status')

,purpose_category,family_status,customer_count,debt_sum,conversion_%
0,операции с автомобилем,Не женат / не замужем,635,82,12.9
1,операции с автомобилем,гражданский брак,430,51,11.9
2,операции с автомобилем,вдовец / вдова,218,20,9.2
3,операции с автомобилем,женат / замужем,2717,226,8.3
4,операции с автомобилем,в разводе,279,21,7.5


Самая высокая доля клиентов с задолженностью с целью `"операции с автомобилем"` -  12.9%. Клиенты в этой категории `не женаты / не замужем`. Далее следует категория клиентов с задолженностью `в гражданском браке` -  11.9%. Самая низкая доля у клиентов `в разводе` - 7.5%.

Рассчитаем долю клиентов с задолженностью с целью `"операции с автомобилем"` в зависимости от `возраста клиентов`:

In [42]:
# выведем на экран результат для цели "операции с автомобилем" и "категории возраста клиентов"
category_conversion(data, 'операции с автомобилем', 'age_category')

,purpose_category,age_category,customer_count,debt_sum,conversion_%
0,операции с автомобилем,от 18 до 25 лет,242,32,13.2
1,операции с автомобилем,от 25 до 30 лет,440,52,11.8
2,операции с автомобилем,от 30 до 40 лет,1155,135,11.7
3,операции с автомобилем,от 40 до 50 лет,1065,91,8.5
4,операции с автомобилем,от 50 лет,1377,90,6.5


Самая высокая доля задолженности с целю `"операции с автомобилем"` у молодых людей в возрасте `от 18 до 25 лет`. Самая низкая доля у клиентов в возрасте `от 50 лет`. Согласно таблице, чем старше клиент - тем ниже доля задолженности.

Рассчитаем долю клиентов с задолженностью с целью `"получение образования"` в зависимости от `уровня дохода`:

In [43]:
# выведем на экран результат для цели "получение образования" и "категории ежемесячного дохода"
category_conversion(data, 'получение образования', 'total_income_category')

,purpose_category,total_income_category,customer_count,debt_sum,conversion_%
0,получение образования,A,4,1,25.0
1,получение образования,C,3014,297,9.9
2,получение образования,B,890,69,7.8
3,получение образования,D,75,2,2.7
4,получение образования,E,5,0,0.0


Самая высокая доля клиентов с задолженностью с целью `"получение образования"` - 25%. Эти клиенты входят в категорию `"A"` с доходом `от 1 000 001`, в которой всего 4 клиента. Далее следует категория клиентов `"C"` с доходом `50 001-200 000` - 9.9%. 

Рассчитаем долю клиентов с задолженностью с целью `"получение образования"` в зависимости от `семейного положения`:

In [44]:
# выведем на экран результат для цели "получение образования" и "категории семейного положения"
category_conversion(data, 'получение образования', 'family_status')

,purpose_category,family_status,customer_count,debt_sum,conversion_%
0,получение образования,гражданский брак,403,60,14.9
1,получение образования,Не женат / не замужем,575,62,10.8
2,получение образования,женат / замужем,2574,215,8.4
3,получение образования,вдовец / вдова,198,15,7.6
4,получение образования,в разводе,238,17,7.1


Самая высокая доля клиентов с задолженностью с целью `"получение образования"` - 14.9%. Клиенты в этой категории состоят в `гражданском браке`. Самую низкую долю имеют клиенты `в разводе` - 7.1%

Рассчитаем долю клиентов с задолженностью с целью `"получение образования"` в зависимости от `возраста клиентов`:

In [45]:
# выведем на экран результат для цели "получение образования" и "категории возраста клиентов"
category_conversion(data, 'получение образования', 'age_category')

,purpose_category,age_category,customer_count,debt_sum,conversion_%
0,получение образования,от 25 до 30 лет,500,67,13.4
1,получение образования,от 18 до 25 лет,215,26,12.1
2,получение образования,от 30 до 40 лет,1047,112,10.7
3,получение образования,от 40 до 50 лет,946,82,8.7
4,получение образования,от 50 лет,1280,82,6.4


Самая высокая доля задолженности с целю `"получение образования"` у клиентов в возрасте `от 25 до 30 лет` - 13.4%. Самая низкая доля у клиентов в возрасте `от 50 лет` - 6.4%. 

Рассчитаем долю клиентов с задолженностью с целью `"проведение свадьбы"` в зависимости от `уровня дохода`:

In [46]:
# выведем на экран результат для цели "проведение свадьбы" и "категории ежемесячного дохода"
category_conversion(data, 'проведение свадьбы', 'total_income_category')

,purpose_category,total_income_category,customer_count,debt_sum,conversion_%
0,проведение свадьбы,C,1729,152,8.8
1,проведение свадьбы,D,34,3,8.8
2,проведение свадьбы,B,544,28,5.1
3,проведение свадьбы,A,2,0,0.0
4,проведение свадьбы,E,4,0,0.0


Клиенты с доходом категории `"С"` - `от 50 001-200 000` и `"D"` - `30 001-50 000` реже остальных оплачивают кредит в срок. Клиенты в категориях `А` - `от 1 000 001` и `Е` - `от 0 - 30 000` совсем не имеют задолженности. В них входит всего 2 и 4 клиентов.

Рассчитаем долю клиентов с задолженностью с целью `"проведение свадьбы"` в зависимости от `семейного положения`:

In [47]:
# выведем на экран результат для цели "проведение свадьбы" и "категории семейного положения"
category_conversion(data, 'проведение свадьбы', 'family_status')

,purpose_category,family_status,customer_count,debt_sum,conversion_%
0,проведение свадьбы,гражданский брак,2313,183,7.9


Все клиенты этой категории находятся в `гражданском браке`.

Рассчитаем долю клиентов с задолженностью с целью `"проведение свадьбы"` в зависимости от `возраста клиентов`:

In [48]:
# выведем на экран результат для цели "проведение свадьбы" и "категории возраста клиентов"
category_conversion(data, 'проведение свадьбы', 'age_category')

,purpose_category,age_category,customer_count,debt_sum,conversion_%
0,проведение свадьбы,от 18 до 25 лет,153,16,10.5
1,проведение свадьбы,от 25 до 30 лет,237,22,9.3
2,проведение свадьбы,от 30 до 40 лет,636,56,8.8
3,проведение свадьбы,от 40 до 50 лет,555,45,8.1
4,проведение свадьбы,от 50 лет,732,44,6.0


Клиенты в возрасте `от 18 до 25 лет` имеют самую большую долю задолженности - 10.5%.

Рассчитаем долю клиентов с задолженностью с целью `"операции с недвижимостью"` в зависимости от `уровня дохода`:

In [49]:
# выведем на экран результат для цели "операции с недвижимостью" и "категории ежемесячного дохода"
category_conversion(data, 'операции с недвижимостью', 'total_income_category')

,purpose_category,total_income_category,customer_count,debt_sum,conversion_%
0,операции с недвижимостью,E,11,2,18.2
1,операции с недвижимостью,C,7997,599,7.5
2,операции с недвижимостью,B,2558,172,6.7
3,операции с недвижимостью,A,17,1,5.9
4,операции с недвижимостью,D,168,6,3.6


Клиенты категории `"E"` c доходом `от 0-30 000` имеют самую высокую долю задолженности - 18.2. В этой группе всего 11 человек, поэтому данные о ней могут исказить результат исследования. Далее следует категория с доходом `от 50 001-200 000` - 7.5%.

Рассчитаем долю клиентов с задолженностью с целью `"операции с недвижимостью"` в зависимости от `семейного положения`:

In [50]:
# выведем на экран результат для цели "операции с недвижимостью" и "категории семейного положения"
category_conversion(data, 'операции с недвижимостью', 'family_status')

,purpose_category,family_status,customer_count,debt_sum,conversion_%
0,операции с недвижимостью,гражданский брак,988,91,9.2
1,операции с недвижимостью,Не женат / не замужем,1586,129,8.1
2,операции с недвижимостью,женат / замужем,6970,486,7.0
3,операции с недвижимостью,в разводе,672,46,6.8
4,операции с недвижимостью,вдовец / вдова,535,28,5.2


Клиенты категории `в гражданском браке` имеют самую большую долю задолженности - 9.2%. Далее следуют клиенты, которые `не женаты / не замужем` - 8.1. Самая низкая доля у клиентов категории `"вдовец / вдова"` - 5.2%.

Рассчитаем долю клиентов с задолженностью с целью `"операции с недвижимостью"` в зависимости от `возраста клиентов`:

In [51]:
# выведем на экран результат для цели "операции с недвижимостью" и "категории возраста клиента"
category_conversion(data, 'операции с недвижимостью', 'age_category')

,purpose_category,age_category,customer_count,debt_sum,conversion_%
0,операции с недвижимостью,от 25 до 30 лет,1295,131,10.1
1,операции с недвижимостью,от 18 до 25 лет,617,57,9.2
2,операции с недвижимостью,от 30 до 40 лет,2855,238,8.3
3,операции с недвижимостью,от 40 до 50 лет,2659,182,6.8
4,операции с недвижимостью,от 50 лет,3325,172,5.2


Клиенты в возрасте `от 25 до 30 лет` имеют задолженность по кредиту чаще, чем клиенты другого возраста - 10.1%. Самая маленькая доля по-прежнему у клиентов в возрасте `от 50 лет` - 5.2%

**Вывод:** 

Согласно данным, разные цели кредита влияют на его возврат в срок в совокупности с такими факторами, как `ежемесячный доход`, `семейное положение` и `возраст клиентов`:

1. Чаще всего не оплачивают кредит в срок клиенты с целью `"операции с автомобилем"` - **9.3%**. Можно предположить, что причинами являются низкий доход клиента и отсутствие финансовой поддержки партнёра. Большинство клиентов с задолженностью с этой целью - молодые люди в возрасте `от 18 до 25 лет`, имеют доход `от 30 001 до 50 000` и входят в категорию `"не женаты / не замужем"`.
2. Клиенты с целью `"получение образования"` не оплачивают кредит своевременно также часто - **9.3%**. Возможные причины - недостаточный уровень дохода из-за необходимости совмещать учёбу с работой, отсутствие финансовой поддержки жены/мужа. Большая часть клиентов - в возрасте `от 25 до 30 лет`, имеет доход `от 50 001-200 000`, состоит в `гражданском браке`. 
3. Меньше доля клиентов с задолженностью с целью `"проведение свадьбы"` - **7.9%**. Чаще её составляют клиенты в возрасте `от 18 до 25 лет`, с доходом `от 50 001-200 000` и `от 30 001-50 000`, состоящие `в гражданском браке`. Возможно, более низкий процент связан с появлением финансовой поддержки жены/мужа после заключения брака.
4. Самая низкая доля задолженности у клиентов с целью `"операции с недвижимостью"` - **7.3%**. Это можно объяснить финансово-грамотным поведением в связи с более зрелым возрастом клиентов, а также наличием серьёзных рисков. Например, возможной потерей жилья при неуплате кредита в срок. Обычно это клиенты в возрасте `от 25 до 30 лет` c доходом `от 50 001-200 000`, состоящие `в гражданском браке`. 

## Общий вывод

Мы исследовали данные, проверили 4 гипотезы и установили:

1. Зависимость между количеством детей и возвратом кредита в срок прослеживается слабо.

Первая гипотеза не подтвердилась. Количество детей в семье практически не влияет на возврат кредита в срок.

2. Существует зависимость между семейным положением клиента и своевременным погашением кредита.

Вторая гипотеза подтвердилась. Особенно заметна разница в 3% между долями в категориях клиентов "не женат / не замужем" и "вдовец / вдова".

3. Из-за отсутствия репрезентативности выборки прослеживается слабая зависимость между уровнем дохода и возвратом кредита в срок. Например, трудно утверждать, что категория клиентов с самым низким доходом - 0-30 000 имеет самый высокий процент задолженности. Вместе с тем самую маленькую долю задолженности составляют клиенты с доходом 30 001-50 000 - 6.0%. 

Таким образом, третья гипотеза не подтвердилась.

4. Можно предположить, что разные цели кредита влияют на погашение кредита в срок следующим образом:
* "операции с автомобилем" - задолженность по причинам низкого дохода, отсутствия финансовой поддержки партнёра, молодого возраста клиентов. Чем младше клиент - тем выше доля задолженности. 
* "получение образования" - неуплата кредита в срок из-за недостаточного уровня дохода, необходимости совмещать учёбу с работой, отсутствия финансовой поддержки жены/мужа.
* "проведение свадьбы" - более низкий процент задолженности в связи с появлением финансовой поддержки после заключения брака.
* "операции с недвижимостью" - оплата в срок в связи с более зрелым возрастом клиентов, наличием серьёзных рисков при операциях с недвижимостью.

Четвёртая гипотеза подтвердилась. Разные цели кредита влияют на его возврат в срок в совокупности с такими факторами, как ежемесячный доход, семейное положение и возраст клиентов.